talks/web3/slides.md

```diff
@@ -0,0 +1,326 @@
+    }
+}
+
+contract Attack {
+    Example e;
+    
+    attack(Example _e) {
+        e = _e;
+        e.cashout(1 ether);
+    }
+    
+    fallback() external payable {
+        e.cashout(1 ether);
+    }
+}
+```
+
+----
+
+### Solidity: Delegate call
+
+```solidity
+contract A {
+    uint test = 0;
+    
+    function c(address b) public {
+        b.delegatecall(
+            abi.encodeWithSignature("stuff()")
+        );
+    }
+}
+
+contract B {
+    uint test = 0; // same as A
+    
+    function stuff() public {
+        // lots of code, expensive to store
+        test = 1337; // A.test = 1337
+    }
+}
+```
+
+----
+
+### Solidity: "randomness"
+
+```solidity
+contract Vuln {
+    function random() public payable returns (uint256) {
+        return uint256(blockhash(block.number-1));
+    }
+}
+```
+
+----
+
+### Solidity: Interfaces
+
+ * ERC: standards & interfaces for smartcontracts
+     * ERC20 popular interface for tokens
+     * ERC721 interface for NFTs
+ * FOSS implementations exists and can be used
+
+----
+
+### Solidity: Real world token
+
+```solidity
+pragma solidity ^0.8.17;
+
+import "https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.0.0/contracts/token/ERC20/ERC20.sol";
+
+contract MyToken is ERC20 {
+    constructor(string memory name, string memory symbol)
+    
+    ERC20(name, symbol) {
+        // Mint 100 tokens to msg.sender
+        // Similar to how
+        // 1 dollar = 100 cents
+        // 1 token = 1 * (10 ** decimals)
+        _mint(msg.sender, 100 * 10 ** uint(decimals()));
+    }
+}
+```
+
+----
+
+### Solidity: ???
+
+ * https://solidity-by-example.org/
+ * https://learnxinyminutes.com/docs/solidity/
+ * https://docs.soliditylang.org
+
+---
+
+### Blockchains in CTFs
+
+ * Transactions cost money: using real chains is expensive
+ * Proof-of-Work costs energy: wastefull
+ * Most CTFs run single node Proof-of-authority chains
+ * Blockchain clients provide client & developer APIs
+     * Only thing that we are interested in
+
+----
+
+### Blockchains in CTFs
+ * Sometimes public testnets are used
+     * Transactions are public: solves are public
+ * Faucets are provided to get coins for free
+ * Sometimes remote provisions contract instance for every player
+
+----
+
+### Tools to use
+
+ * Metamask: wallet, manage accounts
+ * Remix: browser based solidity IDE
+ * web3.js / web3\.py: RPC client libs
+ * hardhat: web3 development framework
+
+----
+
+### Tools: Metamask
+
+ * Browser extension
+ * Can create multiple accounts / keys
+ * Can export keys
+ * Talks to mainnet node by default, supports custom RPCs
+
+----
+
+### Tools: Remix
+
+ * IDE + Debug environment
+ * Usefull to write "exploit contracts"
+ * EVM environment in the browser
+ * Can use "injected metamask" provider to talk to custom chains
+
+----
+
+### Tools: web3.* & hardhat
+
+ * Usefull for more advanced exploits
+ * Provide lots of dev tooling
+
+---
+
+### Challenges
+
+ * Intro chals from Slack
```

Comment:
Could this be improves in some way?
I.e. is the source of the KITCTFCT challenge public or already deployed somewhere?